Oyelowo/surreal-orm
# 9 Enhancement of Data Type Definitions in SurrealDB: Implementation of Comprehensive Data Types and Parse Functions

### Description:

This Pull Request aims to enhance the data type definitions in SurrealDB. It provides extensive support for a variety of data types and implements parsing functionalities to ensure smooth and efficient data handling, making Surreal-orm more versatile and robust.

### Main Features:

1. **Extended Data Types:**
   Surreal-orm now supports a plethora of more explicit/specific data types including but not limited to Geometry Types, Array Types, Set Types, and more.

2. **Parsing Functionality:**
   Implemented parsing functions for various data types ensuring seamless integration and interaction with SurrealDB.

3. **Explicit Data Type Definitions:**
   Enhanced the definition of data types, allowing users to explicitly set the data type of a field in SurrealDB.

4. **Comprehensive Testing:**
   The implementation is thoroughly tested ensuring that each data type and parse function works as intended.

### Detailed Overview:

#### New Data Types:
- `GeometryTypee` supports multiple geometric types like point, line, polygon, and their respective collections.
- `FieldTypee` accommodates various data types including primitive types, record types, geometry types, etc.
- `Array` and `Set` Types support defining the types of items they can store along with the maximum length, with automatic deduplication for Set items.

#### Parsing and Display:
- The implemented parsing functions, `parse_db_field_type` and associated functions, handle the conversion of string inputs to the corresponding data types effectively.
- The Display trait for `GeometryTypee` and `FieldTypee` ensures the correct representation of each data type.

#### Utility Functions:
- Utility functions are provided to check the kind of a field type, like `is_primitive`, `is_record`, `is_geometry`, etc., facilitating more informed and precise data handling.

#### Testing:
- Comprehensive testing is done for each data type and parse function, ensuring their correctness and reliability.

### Example:
```rust
use surrealdb::{FieldTypee, parse_db_field_type};
assert_eq!(parse_db_field_type("any"), Ok(("", FieldTypee::Any)));
assert_eq!(parse_db_field_type("null"), Ok(("", FieldTypee::Null)));
assert_eq!(parse_db_field_type("int | option<float> | array<option<string>|int|null, 10> | set<option<number>|float|null, 10> | option<array> | option<set<option<int>>>"),
        FieldTypee::Union(vec![
            FieldTypee::Int,
            FieldTypee::Option(Box::new(FieldTypee::Float)),
            FieldTypee::Array(
                Box::new(FieldTypee::Union(vec![
                    FieldTypee::Option(Box::new(FieldTypee::String)),
                    FieldTypee::Int,
                    FieldTypee::Null
                ])),
                Some(10)
            ),
            FieldTypee::Set(
                Box::new(FieldTypee::Union(vec![
                    FieldTypee::Option(Box::new(FieldTypee::Number)),
                    FieldTypee::Float,
                    FieldTypee::Null
                ])),
                Some(10)
            ),
            FieldTypee::Option(Box::new(FieldTypee::Array(
                Box::new(FieldTypee::Any),
                None
            ))),
            FieldTypee::Option(Box::new(FieldTypee::Set(
                Box::new(FieldTypee::Option(Box::new(FieldTypee::Int))),
                None
            )))
        ])
    );
```

### Impact:
This enhancement will significantly impact the versatility of surreal-orm, allowing users to handle various data types efficiently, and ensuring more robust and flexible data management.

### Future Work:
- Further optimization/refactoring of parse functions.

### How to Test:
- Ensure that you have the latest version of SurrealDB and all its dependencies installed.
- Run the provided test cases in your local environment to verify the correctness of the new data types and parsing functions.

### Reviewers:
- N/A

### Notes to Reviewers:
- Please pay close attention to the parsing functions and their handling of different data types.
- Verify the implementation against the provided test cases and ensure all edge cases are covered.

---

### Changes:
- Added new data types in `FieldTypee` and `GeometryTypee`.
- Implemented parsing functions.
- Provided display traits for newly added data types.
- Added comprehensive test cases for each new data type and parse function.

### Files Changed:
- `field_type_new.rs` added

### Checklist:

### Acknowledgements:
-

### Screenshots:
- N/A
